File: server/routes/categories.js
const express = require('express');
const { v4: uuidv4 } = require('uuid');
const db = require('../config/db');
const { authenticate, authorize } = require('../middleware/auth');

const router = express.Router();

// GET /api/categories - List all active categories (public)
router.get('/', async (req, res) => {
  try {
    const categories = await db.prepare('SELECT * FROM categories WHERE active = 1 ORDER BY sort_order ASC, name ASC').all();
    res.json({ categories });
  } catch (err) {
    console.error('Error listing categories:', err);
    res.status(500).json({ error: 'Failed to list categories' });
  }
});

// GET /api/categories/all - List all categories including inactive (admin only)
router.get('/all', authenticate, authorize('admin'), async (req, res) => {
  try {
    const categories = await db.prepare('SELECT * FROM categories ORDER BY sort_order ASC, name ASC').all();
    res.json({ categories });
  } catch (err) {
    console.error('Error listing all categories:', err);
    res.status(500).json({ error: 'Failed to list categories' });
  }
});

// POST /api/categories - Create a new category (admin only)
router.post('/', authenticate, authorize('admin'), async (req, res) => {
  try {
    const { name, icon, description, sortOrder } = req.body;
    if (!name || !name.trim()) {
      return res.status(400).json({ error: 'Category name is required' });
    }

    // Check for duplicate
    const existing = await db.prepare('SELECT id FROM categories WHERE name = ?').get(name.trim());
    if (existing) {
      return res.status(409).json({ error: 'Category with this name already exists' });
    }

    const id = uuidv4();
    await db.prepare('INSERT INTO categories (id, name, icon, description, sort_order) VALUES (?, ?, ?, ?, ?)')
      .run(id, name.trim(), icon || '📋', description || '', sortOrder || 0);

    const category = await db.prepare('SELECT * FROM categories WHERE id = ?').get(id);
    res.status(201).json({ category });
  } catch (err) {
    console.error('Error creating category:', err);
    res.status(500).json({ error: 'Failed to create category' });
  }
});

// PUT /api/categories/:id - Update a category (admin only)
router.put('/:id', authenticate, authorize('admin'), async (req, res) => {
  try {
    const category = await db.prepare('SELECT * FROM categories WHERE id = ?').get(req.params.id);
    if (!category) {
      return res.status(404).json({ error: 'Category not found' });
    }

    const { name, icon, description, sortOrder, active } = req.body;

    // Check for duplicate name if name changed
    if (name && name.trim() !== category.name) {
      const existing = await db.prepare('SELECT id FROM categories WHERE name = ? AND id != ?').get(name.trim(), req.params.id);
      if (existing) {
        return res.status(409).json({ error: 'Category with this name already exists' });
      }
    }

    await db.prepare(`UPDATE categories SET 
      name = COALESCE(?, name), 
      icon = COALESCE(?, icon), 
      description = COALESCE(?, description), 
      sort_order = COALESCE(?, sort_order),
      active = COALESCE(?, active),
      updated_at = CURRENT_TIMESTAMP 
      WHERE id = ?`)
      .run(
        name?.trim() || null,
        icon || null,
        description || null,
        sortOrder !== undefined ? sortOrder : null,
        active !== undefined ? (active ? 1 : 0) : null,
        req.params.id
      );

    const updated = await db.prepare('SELECT * FROM categories WHERE id = ?').get(req.params.id);
    res.json({ category: updated });
  } catch (err) {
    console.error('Error updating category:', err);
    res.status(500).json({ error: 'Failed to update category' });
  }
});

// DELETE /api/categories/:id - Delete a category (admin only)
router.delete('/:id', authenticate, authorize('admin'), async (req, res) => {
  try {
    const category = await db.prepare('SELECT * FROM categories WHERE id = ?').get(req.params.id);
    if (!category) {
      return res.status(404).json({ error: 'Category not found' });
    }

    // Instead of deleting, set active to 0 (soft delete)
    await db.prepare('UPDATE categories SET active = 0, updated_at = CURRENT_TIMESTAMP WHERE id = ?').run(req.params.id);
    res.json({ success: true, message: 'Category deactivated' });
  } catch (err) {
    console.error('Error deleting category:', err);
    res.status(500).json({ error: 'Failed to delete category' });
  }
});

module.exports = router;

File: python_backend/routers/categories.py
"""Category routes: CRUD for service categories."""

import uuid
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from python_backend.database import get_db
from python_backend.auth import get_current_user, require_role
from python_backend.schemas import CategoryCreateRequest, CategoryUpdateRequest
from python_backend.serializers import row, rows
from python_backend.cache import cache_ttl

router = APIRouter(prefix="/api/categories", tags=["categories"])


@router.get("")
@cache_ttl(seconds=120)
async def list_categories(active: str = None, db: AsyncSession = Depends(get_db)):
    query = "SELECT * FROM categories WHERE 1=1"
    params = {}
    if active == "true":
        query += " AND active = 1"
    query += " ORDER BY sort_order ASC"
    result = await db.execute(text(query), params)
    return {"categories": rows(result.mappings().all())}


@router.get("/all")
@cache_ttl(seconds=120)
async def list_all(db: AsyncSession = Depends(get_db)):
    result = await db.execute(text("SELECT * FROM categories ORDER BY sort_order ASC"))
    return {"categories": rows(result.mappings().all())}


@router.post("")
async def create_category(req: CategoryCreateRequest, user: dict = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    if user["role"] != "admin":
        raise HTTPException(403, "Unauthorized")
    cat_id = str(uuid.uuid4())
    await db.execute(
        text("INSERT INTO categories (id, name, icon, description, sort_order) VALUES (:id, :name, :icon, :desc, :so)"),
        {"id": cat_id, "name": req.name, "icon": req.icon, "desc": req.description, "so": req.sortOrder},
    )
    result = await db.execute(text("SELECT * FROM categories WHERE id = :id"), {"id": cat_id})
    return {"category": row(result.mappings().first())}


@router.put("/{cat_id}")
async def update_category(cat_id: str, req: CategoryUpdateRequest, user: dict = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    if user["role"] != "admin":
        raise HTTPException(403, "Unauthorized")
    await db.execute(
        text("""
            UPDATE categories SET
                name = COALESCE(:name, name),
                icon = COALESCE(:icon, icon),
                description = COALESCE(:desc, description),
                sort_order = COALESCE(:so, sort_order),
                active = COALESCE(:active, active),
                updated_at = CURRENT_TIMESTAMP
            WHERE id = :id
        """),
        {"name": req.name, "icon": req.icon, "desc": req.description, "so": req.sortOrder, "active": req.active, "id": cat_id},
    )
    result = await db.execute(text("SELECT * FROM categories WHERE id = :id"), {"id": cat_id})
    return {"category": row(result.mappings().first())}


@router.delete("/{cat_id}")
async def delete_category(cat_id: str, user: dict = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    if user["role"] != "admin":
        raise HTTPException(403, "Unauthorized")
    await db.execute(text("DELETE FROM categories WHERE id = :id"), {"id": cat_id})
    return {"message": "Category deleted"}
